### DISCUSSION TITLE
How to disable a single sink?

### DISCUSSION THREAD

**Discussion by CJCombrink** [gabime/spdlog#2843] [Q&A] [answered]
I have a default logger with multiple sinks: `stdout`, `stderr`, `syslog`

For a specific part/path of the application I need to dump output (JSON) to the console. For that path I need to disable the `stdout` sink.
This path is command line argument driven (`--dump`)

I use the default logger on startup and add the rest of the sinks. 
My class then calls `auto logger = spdlog::default_logger()->clone("my_context");` 
What I then need to do is (pseudo code)
```
for all stdout sinks in logger:
   set_level(spdlog::level::off);
```

I want to clone the default since I still need to use the other sinks without having to recreate any of them. 

Is this possible or should I rethink my flow?
Any help or suggestions are welcome. 

PS: The commands are dynamic. At the time where the default logger is set up, the application does not know that `--dump` is an option.

**Comment by tt4g** [comment]
If you just want to implement pseudo code, it looks like this:

```cpp
for (auto sink : logger->sinks()) {
    std::shared_ptr<SINK_T> stdout_sink = std::dynamic_pointer_cast<SINK_T>(sink);
    if (stdout_sink) {
        stdout_sink->set_level(spdlog::level::off);
    }
}
```

However, the logging level must be restored.

**Reply by gabime** [reply, marked as answer]
Perhaps just remove those sinks instead of setting the level. This way you don't need to restore the levels (and performance is better since the new logger wont store those sinks at all).

**Reply by gabime** [reply]
Also note that iterating over the sinks vector is not thread safe.See the wiki for [details](https://github.com/gabime/spdlog/wiki/1.1.-Thread-Safety#non-thread-safe-functions).

**Reply by CJCombrink** [reply]
Fortunately for this use there is only the main thread.

**Comment by CJCombrink** [comment]
I decided to go with the following, mostly for the reason that I don't like dynamic cast:
```
const std::string LOGGER_STDOUT{"stdout"};

auto stdout_sink = std::make_shared<spdlog::sinks::stdout_color_sink_mt>();
auto cerr_sink   = std::make_shared<spdlog::sinks::stderr_color_sink_mt>();
... // other sinks

std::vector<spdlog::sink_ptr> sinks;
sinks.push_back(stdout_sink);
sinks.push_back(...);

auto combined_logger = std::make_shared<spdlog::logger>(std::string{}, begin(sinks), end(sinks));
spdlog::set_default_logger(combined_logger);

auto stdout_logger = std::make_shared<spdlog::logger>(LOGGER_STDOUT, stdout_sink);
spdlog::register_logger(stdout_logger);

// Elsewhere for dump:
std::ranges::for_each(spdlog::get(LOGGER_STDOUT)->sinks(), [](auto sink) {
    sink->set_level(spdlog::level::off);
});
```

But I will consider reworking it to remove the sink instead.
